### PR TITLE
test(mme): Network initiated bearer activation and deactivation

### PR DESCRIPTION
**Pull request by ulaskozat**
Signed-off-by: Ulas Kozat <kozat@fb.com>

## Summary

- Added unit test for network initiated bearer activation and deactivation.
- Fixed issues in counting s1u bearers.
- Fixed an enumeration bug

## Test Plan

`make test_oai`
## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.c [hunk 1]
```diff
@@ -5302,7 +5302,7 @@ status_code_e s1ap_mme_handle_erab_rel_response(
   }
   S1AP_E_RAB_REL_RSP(message_p).mme_ue_s1ap_id = ue_ref_p->mme_ue_s1ap_id;
   S1AP_E_RAB_REL_RSP(message_p).enb_ue_s1ap_id = ue_ref_p->enb_ue_s1ap_id;
-  S1AP_E_RAB_REL_RSP(message_p).e_rab_rel_list.no_of_items           = 1;
+  S1AP_E_RAB_REL_RSP(message_p).e_rab_rel_list.no_of_items           = 0;
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/s1ap/s1ap_mme_handlers.c:5355]: 👍

**Review by ardzoht — APPROVED** [review]
lgtm

File: lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp [hunk 2]
```diff
@@ -1546,5 +1547,144 @@ TEST_F(MmeAppProcedureTest, TestAttachIdleServiceReqDetach) {
...
+
+  // Check MME state after Modify Bearer Response
+  send_activate_message_to_mme_app();
+  cv.wait_for(lock, std::chrono::milliseconds(STATE_MAX_WAIT_MS));
+  EXPECT_EQ(mme_state_p->nb_ue_attached, 1);
+  EXPECT_EQ(mme_state_p->nb_ue_connected, 1);
+  EXPECT_EQ(mme_state_p->nb_default_eps_bearers, 1);
+  EXPECT_EQ(mme_state_p->nb_ue_idle, 0);
+
+  // Send activate dedicated bearer request mimicing S1AP
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp]:
```suggestion
  // Send activate dedicated bearer request mimicing S11
```

File: lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp [hunk 3]
```diff
@@ -1546,5 +1547,144 @@ TEST_F(MmeAppProcedureTest, TestAttachIdleServiceReqDetach) {
...
+  // Check MME state after Bearer Activation
+  send_activate_message_to_mme_app();
+  cv.wait_for(lock, std::chrono::milliseconds(STATE_MAX_WAIT_MS));
+  EXPECT_EQ(mme_state_p->nb_ue_attached, 1);
+  EXPECT_EQ(mme_state_p->nb_ue_connected, 1);
+  EXPECT_EQ(mme_state_p->nb_default_eps_bearers, 1);
+  EXPECT_EQ(mme_state_p->nb_s1u_bearers, 2);
+  EXPECT_EQ(mme_state_p->nb_ue_idle, 0);
+
+  uint8_t ebi_to_be_deactivated = 6;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp:1639]: Ebi is 5 for default bearer
```suggestion
  uint8_t ebi_to_be_deactivated = 5;
```

File: lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp [hunk 4]
```diff
@@ -1546,5 +1547,144 @@ TEST_F(MmeAppProcedureTest, TestAttachIdleServiceReqDetach) {
...
+  // Constructing and sending deactivate bearer request
+  // for default bearer that should trigger session termination
+  send_s11_deactivate_bearer_req(1, &ebi_to_be_deactivated, false);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp:1642]: Would this session termination trigger Implicit Detach for UE?

File: lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp (hunk 3 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp:1639]: this is deactivating dedicated bearer not the default.

File: lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp (hunk 4 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp:1642]: this test is activating and deactivating dedicated bearer. Default bearer deactivation test is done as part of network initiated deactivation.

Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp:1642]: I will make the test name more explicit for dedicated bearer.

Comment (ssanadhya) [reply lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp:1642]: I see, then the comment in lines 1638-1639 should be changed

Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp:1642]: yes indeed, good catch...

File: lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp (hunk 2 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp]: good catch...

**Pull request merged**